erpnext/accounts/doctype/pos_invoice/pos_invoice.py

```diff
@@ -212,6 +213,25 @@ def validate_serialised_or_batched_item(self):
 		if error_msg:
 			frappe.throw(error_msg, title=_("Invalid Item"), as_list=True)
 
+	def validate_serial_nos(self):
+		error_msg = []
+		for item in self.get("items"):
+			serialized = item.get("has_serial_no")
+			msg = ""
+			if serialized:
+				valid_serial_nos = frappe.get_all('Serial No', filters={'item_code':item.get("item_code")}, pluck="name")
+				invalid_serials = ""
+				for serial_no in item.get('serial_no').split("\n"):
+					if serial_no not in valid_serial_nos:
```

Comment:
You can do basic set difference here instead. 

invalid_serial_nos = serial_no_in_doc - `real_serial_no`

Reply:
I felt the loop could be simpler and helps with better error message

Reply:
You can get the same error message. Anyway, doesn't matter much.